에픽 관련 소켓로직 작성 및 로드맵 컴포넌트 구현

rudy3091 opened the pull request:
## 😀 제목

에픽 관련 소켓로직 작성 및 로드맵 컴포넌트 구현

## ⛅️ 내용

> 이 PR의 작업 요약

- useSocketReceive 관련 에러를 해결했습니다. 00cd2a316c51be1b7c6dd001c7b3ec088cbbcf22
- 로드맵 컴포넌트가 내용물을 scroll로 보여주는게 아니라 컨테이너 사이즈를 변경해 보여주도록 수정했습니다. 67b1ea7a87bf3dcb317e5985e4e90e6acf33ff89
- 소켓 이벤트 발행 및 구독 로직을 추가했습니다.
- 로드맵 컴포넌트에서 RoadmapCalendar 컴포넌트를 따로 분리했습니다.

## 🎸특이사항

> 리뷰시 참고할만한 내용, 주의깊게 봐줬으면 하는 내용

에러 핸들링을 제대로 했는지 고민이네요

우선은 최대한 에러케이스를 고려해서 toast로 사용자에게 시각적인 피드백을 제공하도록 했습니다

이 부분 살펴봐주세용

rudy3091 committed:
fix: 중복되는 변수이름 삭제 [#18]

중복되는 epic과 epicsOnProject 중 전자는 제거하고 후자만 남김

rudy3091 committed:
fix: 에픽 생성 submithandler 수정 [#103]

에픽 생성시 input의 submit handler 함수가 비동기적으로 동작하도록 수정

rudy3091 committed:
feat: 에픽 생성시 소켓이벤트를 발생시키도록 수정 [#103]

- 에픽을 생성하면 소켓 이벤트 "NEW_EPIC"가 발생하도록 구현
- EpicPlaceholder 컴포넌트에 전달하는 handleSubmit 함수를 가독성 좋게
  밖으로 뺌

rudy3091 committed:
feat: 새로운 에픽 생성 소켓이벤트 발생 시 broadcast [#103]

- 새로운 에픽 항목이 생성되면 서버에서는 이를 감지해 에픽을 생성한
  클라이언트 외의 모든 클라이언트들에게 이를 broadcast함
- 클라이언트들은 이를 감지해 새로운 요청을 보내도록 해야함

rudy3091 committed:
feat: workpage 에서 소켓 이벤트 구독 [#103]

다른 클라이언트에서 에픽을 추가할 시 발생하는 소켓 이벤트를workpage에서
구독함
현재 페이지가 바뀔 때마다 이벤트 핸들러가 중복으로 등록돼서 수정이
필요함

rudy3091 committed:
refactor: dead code 제거

workpage 에서 사용하지 않는 type 정의 제거

rudy3091 committed:
docs: 소켓이벤트 발행 관련 주석 추가

소켓 필터링 관련 메모
구현 완료시 삭제 예정

rudy3091 committed:
fix: socket receive hook 문제 해결

socket.on 메소드의 콜백 함수에 이벤트핸들러 함수를 직접 전달해 컴포넌트
언마운트시 이벤트핸들러를 해제하도록 수정

rudy3091 committed:
refactor: cors옵션 제거

서버 app단에서 cors를 이미 제거했기 때문에 라우터단의 cors 설정을 제거함

rudy3091 committed:
feat: getEpicById 함수 추가 [#103]

새로운 에픽이 생성되면 이 함수를 이용해 그 에픽에 대한 정보를 조회함

rudy3091 committed:
fix: 에픽 조회 로직 문제 해결 [#103]

에픽을 조회할 떄 get 메소드로 요청하도록 수정
data의 타입을 epic 객체에 맞게 수정

rudy3091 committed:
feat: 에픽생성 핸들러 수정 [#103]

에픽 생성시 실행할 이벤트 핸들러에 toast 를 통한 피드백 대신 새롭게
생성한 에픽의 id를 기준으로 api를 호출
응답결과를 context 객체에 append해서 데이터 동기화

rudy3091 committed:
layout: 로드맵 컴포넌트의 레이아웃 수정 [#18]

로드맵 컴포넌트 컨테이너에 overflow 대신 내용물에 따라 늘어나는
레이아웃을 적용함

rudy3091 committed:
feat: 로드맵 내부 calendar 컴포넌트 분리 [#18]

Roadmap 컴포넌트 내부의 오른쪽 영역에 해당하는 영역을 별도의 컴포넌트로
분리

rudy3091 committed:
feat: RoadmapCalendar 컴포넌트 스타일 추가 [#18]

간단한 ui 코드 작성

rudy3091 committed:
layout: RoadmapCalendar 컴포넌트 스타일 추가 [#18]

Calendar 헤더 영역에 일자를 바꿀 수 있는 버튼 추가

rudy3091 committed:
layout: 버튼 색상 변경 [#18]

프로토타입 디자인에 없는 부분이라 조금 더 수정할 예정

rudy3091 committed:
refactor: Dead code 제거

사용하지 않는 모듈 import 제거

ChanHoLee275 commented on client/src/components/RoadmapCalendar/index.tsx:
```diff
@@ -0,0 +1,25 @@
+        </S.ButtonWrapper>
+      </S.CalendarHead>
+      <S.DaysWrapper>
+        {Array.from(new Array(14), (_, i) => (
+          // TODO: key props를 index 값이 아닌 date 값과 연관한 값으로 변경
+          <div key={i}>{i + 1}</div>
+        ))}
```
작성하신 코드도 좋아보이지만
```typescript
[...Array(14)].map((el, i) => {<div key={i}>{i+1}</div>})
```
이렇게 작성하는 것은 어떨까요..?

ChanHoLee275 replied:
취향 차이이긴 합니다 ㅎㅎ

ChanHoLee275 commented on client/src/layers/Roadmap/index.tsx:
```diff
@@ -50,7 +57,7 @@ const Roadmap = ({ projectId }: RoadmapProps) => {
           </Button>
         </S.EpicEntry>
-        <S.RoadmapCalendar>content</S.RoadmapCalendar>
+        <RoadmapCalendar></RoadmapCalendar>
```
아무 children도 내려주지 않으면 아래와 같이 바꾸는 것도 좋아보입니다!
```typescript
<RoadmapCalendar/>
```

ChanHoLee275 commented on client/src/components/EpicPlaceholder/index.tsx:
```diff
@@ -9,9 +9,9 @@ interface EpicPlaceholderProps {
 
-  const handleEnter = (ev: React.KeyboardEvent) => {
+  const handleEnter = async (ev: React.KeyboardEvent) => {
     if (value === '' || ev.key !== 'Enter') return;
-    handleSubmit(value);
+    await handleSubmit(value);
```
handleSubmit의 값을 사용하지 않는데, 비동기로 처리하신 이유가 있을까요?

haesoo-y commented on client/src/components/EpicPlaceholder/index.tsx:
```diff
@@ -9,9 +9,9 @@ interface EpicPlaceholderProps {
 
-  const handleEnter = (ev: React.KeyboardEvent) => {
+  const handleEnter = async (ev: React.KeyboardEvent) => {
     if (value === '' || ev.key !== 'Enter') return;
```
`S.Container`를 form태그로 두시면 input에 텍스트를 입력후 엔터치면 form 태그의 onSubmit이 발동할텐데 굳이 input에 onKeyPress를 쓰고 조건문을 사용해서 enter을 구분하실 필요가 있으실까요??

haesoo-y commented on client/src/layers/Roadmap/index.tsx:
```diff
@@ -1,20 +1,44 @@
+          id: id,
+          name: value,
+          startAt: new Date(),
+          endAt: new Date(),
```
아까 찬호님이랑도 얘기 나눴었는데, 서버에서 받아온 json 에는 date형식이 string으로 되어있지 않나요?? 그럼 이런식으로 넣게되면 epic에 어떤 객체는 date 객체로 날짜를 갖고 어떤 객체는 string으로 날짜를 갖게 될 것 같은데 어떻게 생각하시나요?? `LOAD_EPIC`시 map으로 받아온 날짜들을 데이트 객체로 바꿔야 할까요??

haesoo-y commented on client/src/layers/Roadmap/index.tsx:
```diff
@@ -1,20 +1,44 @@
+
+  const handleSubmit = async (value: string) => {
+    try {
+      if (!projectId) throw new Error('유저 정보 없음');
```
에러 메세지를 다른 곳은 ~~습니다로 통일 시켰는데 여기는 명사로 끝내셨네요 ㅎㅎ

haesoo-y reviewed (commented):
에러를 전체적으로 꼼꼼하게 처리하셨네요. 좋습니다 😃

ingong commented on client/src/lib/api/epic.ts:
```diff
@@ -34,3 +34,21 @@ export const createEpic = async (projectId: number | string, epicName: string) =
+  try {
+    const result: {
+      code: number;
+      data: { id: number; name: string; startAt: Date; endAt: Date };
```
이 부분은 Type 을 가져와서 할 수 있지않을까요?

ingong commented on client/src/layers/Roadmap/index.tsx:
```diff
@@ -1,20 +1,44 @@
-import Button from '@/lib/design/Button';
+import { toast } from 'react-toastify';
 import S from './style';
 import EpicPlaceholder from '../../components/EpicPlaceholder';
```
절대경로로 해도 좋을 것 같네요...ㅎ

ingong commented on client/src/layers/Roadmap/index.tsx:
```diff
@@ -1,20 +1,44 @@
 import EpicPlaceholder from '../../components/EpicPlaceholder';
 import { useEpicDispatch, useEpicState } from '@/lib/hooks/useContextHooks';
 import { createEpic } from '@/lib/api/epic';
+import useSocketSend from '@/lib/hooks/useSocketSend';
```
지금 hooks 를 두 번 import 하게 되는데,
index 파일을 만들고 default 로 export 하는 식으로 하는거 고려해보는거 어떨까요?

ingong reviewed (commented):
잘 읽고 갑니다..!

rudy3091 committed:
chore: yarn.lock 파일 추가

누락된 lockfile 변경사항 반영

rudy3091 committed:
fix: 에픽 입력시 form 태그를 사용하도록 수정

엔터키를 판별할필요없이 그냥 form 태그를 이용해 submit 이벤트를 처리

rudy3091 committed:
fix: 필요없는 async await 로직 제거 [#103]

input 내부 value를 클리어해주는 것과 에픽 생성 요청을 처리하는건 따로
관리해도 무방함

rudy3091 committed:
refactor: lib/hooks/index.ts 에서 모든 훅을 export

custom hook 함수들을 한 파일에서 import 할 수 있도록 lib/hooks/index.ts
에서 모두 export해줌

rudy3091 committed:
refactor: 로드맵 컴포넌트 리팩토링 [#103]

epicDispatcher를 사용하는 과정에서 action 객체를 생성하는 함수를 외부로
빼내 좀 더 가독성있게 작성

rudy3091 committed:
refactor: 에러 핸들링 로직 변경 [#103]

- epic 생성에 실패하면 id = -1 을 반환하는것이 아니라 실패했을때 바로
  toast 알림을 띄움
- 반환한 id를 받는 쪽에서 id를 검사하는것이 아니라 result 자체가 없다면
  에러를 발생시킴

rudy3091 committed:
fix: epic api 호출 에러관련 주석 추가 [#103]

변경한 에러 핸들링 로직에 따른 주석 추가

rudy3091 committed:
feat: toast 에 들어갈 메시지들을 상수화

특정 작업 성공시, 실패시 보여줄 메시지들을 상수화해 파일 하나에 모아놓음

rudy3091 committed:
refactor: 에러메시지 상수화 적용

상수화한 에러 메시지들을 api 호출 함수들에 적용함

rudy3091 committed:
chore: 에러핸들링 방식에 관한 주석 추가

해결되면 제거할것

rudy3091 committed:
refactor: 에러메시지 및 성공메시지 상수화 적용

컴포넌트 내부 에러 핸들링 로직에 에러 및 성공 메시지 상수화 적용

rudy3091 committed:
refactor: 필드 일부가 문자열 타입인 에픽 타입 추가

startAt, endAt 필드가 string 타입인 EpicWithString 타입 추가
서버에서 보내주는 데이터에는 Date 타입이 문자열로 저장되어있기에 추가

rudy3091 committed:
refactor: 유틸함수 추가

에픽 객체의 문자열 날짜를 Date타입으로 변환시켜줄 함수

rudy3091 committed:
refactor: 에픽 context reducer 함수 수정

처음 서버에서 에픽 데이터들을 받아올때 추가할때 startAt 과 endAt 은
string 타입이므로 변경함

rudy3091 committed:
refactor: 에픽 타입에 타입가드 추가

로직에 대한 코드가 들어갔으므로 파일타입을 .d.ts 에서 .ts로 변경

rudy3091 committed:
refactor: 타입가드를 사용해 에픽 context 로직 수정

서버에서 받아온 데이터라면 EpicWithString 타입이므로 이를 타입가드를 통해
판별한 뒤 변환과정을 추가함

rudy3091 committed:
refactor: 코드스타일 변경

children 이 없는 JSX 엘리먼트에는 self closing tag 사용

rudy3091 committed:
refactor: 타이핑 코드 스타일 수정

이미 선언된 타입 정의를 사용해 코드 간략화

rudy3091 committed:
fix: merge conflict 해결
